File: Heap/K_Closest_Points_to_Origin.py
#Problem Link:- https://leetcode.com/problems/k-closest-points-to-origin/

"""We have a list of points on the plane.  Find the K closest points to the origin 
(0, 0).(Here, the distance between two points on a plane is the Euclidean distance.)
You may return the answer in any order.  The answer is guaranteed to be unique (except 
for the order that it is in.)"""

class Solution:
    def kClosest(self, points: List[List[int]], K: int) -> List[List[int]]:
        import math
        import heapq
        n = len(points)
        if n == 1:
            return points[0]
        heap = []
        ans = []
        for i in range(n):
            x = points[i][0]
            y = points[i][1]
            dist = math.sqrt(x*x + y*y)
            heapq.heappush(heap,(dist,i))
        for i in range(K):
            d,ind = heapq.heappop(heap)
            ans.append(points[ind])
        return ans

File: Heap/KClosestPointstoOrigin.cpp
//Problem Link:- https://leetcode.com/problems/k-closest-points-to-origin/

#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        priority_queue<pair<double,int>,vector<pair<double,int>>,greater<pair<double,int>>> pq;
        for(int i=0;i<points.size();i++){
            int x = points[i][0];
            int y = points[i][1];
            int dist = x*x + y*y;
            double result = sqrt(dist);
            pq.push({result,i});
        }
        vector<vector<int>> result;
        for(int i=0;i<k;i++){
            int idx = pq.top().second;
            pq.pop();
            result.push_back(points[idx]);
        }
        return result;
    }
};